`TestQueryLog` CI test flaky

### What did you do?

Going about my day reviewing PRs, watching CI tests fail

### What did you expect to see?

No CI test failures

### What did you see instead? Under which circumstances?

The `TestQueryLog` test failed in CI, seems flaky:

```
--- FAIL: TestQueryLog (52.62s)
    --- FAIL: TestQueryLog/rule_queries,_127.0.0.1:34361 (10.52s)
        query_log_test.go:356: 
            	Error Trace:	/__w/prometheus/prometheus/cmd/prometheus/query_log_test.go:356
            	            				/__w/prometheus/prometheus/cmd/prometheus/query_log_test.go:427
            	Error:      	"6" is not greater than "6"
            	Test:       	TestQueryLog/rule_queries,_127.0.0.1:34361
            	Messages:   	no queries logged
        query_log_test.go:273: ts=2024-10-02T00:51:24.581Z caller=main.go:601 level=info msg="No time or size retention was set so using the default time retention" duration=15d
            ts=2024-10-02T00:51:24.581Z caller=main.go:645 level=info msg="Starting Prometheus Server" mode=server version="(version=, branch=, revision=unknown)"
            ts=2024-10-02T00:51:24.581Z caller=main.go:650 level=info build_context="(go=go1.23.1, platform=linux/amd64, user=, date=, tags=dedupelabels)"
            ts=2024-10-02T00:51:24.581Z caller=main.go:651 level=info host_details="(Linux 6.8.0-1014-azure #16~22.04.1-Ubuntu SMP Thu Aug 15 21:31:41 UTC 2024 x86_64 10bad815c49d (none))"
            ts=2024-10-02T00:51:24.581Z caller=main.go:652 level=info fd_limits="(soft=1048576, hard=1048576)"
            ts=2024-10-02T00:51:24.581Z caller=main.go:653 level=info vm_limits="(soft=unlimited, hard=unlimited)"
            ts=2024-10-02T00:51:24.583Z caller=web.go:651 level=info component=web msg="Start listening for connections" address=127.0.0.1:34361
            ts=2024-10-02T00:51:24.584Z caller=main.go:1192 level=info msg="Starting TSDB ..."
            ts=2024-10-02T00:51:24.586Z caller=tls_config.go:348 level=info component=web msg="Listening on" address=127.0.0.1:34361
            ts=2024-10-02T00:51:24.586Z caller=tls_config.go:351 level=info component=web msg="TLS is disabled." http2=false address=127.0.0.1:34361
            ts=2024-10-02T00:51:24.588Z caller=head.go:632 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
            ts=2024-10-02T00:51:24.588Z caller=head.go:719 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=961ns
            ts=2024-10-02T00:51:24.588Z caller=head.go:727 level=info component=tsdb msg="Replaying WAL, this may take a while"
            ts=2024-10-02T00:51:24.588Z caller=head.go:799 level=info component=tsdb msg="WAL segment loaded" segment=0 maxSegment=0
            ts=2024-10-02T00:51:24.588Z caller=head.go:836 level=info component=tsdb msg="WAL replay completed" checkpoint_replay_duration=64.03µs wal_replay_duration=143.178µs wbl_replay_duration=140ns chunk_snapshot_load_duration=0s mmap_chunk_replay_duration=961ns total_replay_duration=236.293µs
            ts=2024-10-02T00:51:24.589Z caller=main.go:1213 level=info fs_type=794c7630
            ts=2024-10-02T00:51:24.589Z caller=main.go:1216 level=info msg="TSDB started"
            ts=2024-10-02T00:51:24.589Z caller=main.go:1399 level=info msg="Loading configuration file" filename=/tmp/config654002981
            ts=2024-10-02T00:51:24.590Z caller=main.go:1438 level=info msg="updated GOGC" old=100 new=75
            ts=2024-10-02T00:51:24.590Z caller=main.go:1449 level=info msg="Completed loading of configuration file" filename=/tmp/config654002981 totalDuration=521.97µs db_storage=1.052µs remote_storage=2.024µs web_handler=331ns query_engine=711ns scrape=159.329µs scrape_sd=902ns notify=741ns notify_sd=501ns rules=199.664µs tracing=5.21µs
            ts=2024-10-02T00:51:24.590Z caller=main.go:1177 level=info msg="Server is ready to receive web requests."
            ts=2024-10-02T00:51:24.590Z caller=manager.go:164 level=info component="rule manager" msg="Starting rule manager..."
            ts=2024-10-02T00:51:27.061Z caller=main.go:1399 level=info msg="Loading configuration file" filename=/tmp/config654002981
            ts=2024-10-02T00:51:27.062Z caller=main.go:1449 level=info msg="Completed loading of configuration file" filename=/tmp/config654002981 totalDuration=544.171µs db_storage=1.353µs remote_storage=1.483µs web_handler=512ns query_engine=17.863µs scrape=67.637µs scrape_sd=1.473µs notify=931ns notify_sd=851ns rules=169.438µs tracing=5.149µs
            ts=2024-10-02T00:51:29.063Z caller=main.go:1399 level=info msg="Loading configuration file" filename=/tmp/config654002981
            ts=2024-10-02T00:51:29.064Z caller=main.go:1449 level=info msg="Completed loading of configuration file" filename=/tmp/config654002981 totalDuration=489.689µs db_storage=1.523µs remote_storage=2.144µs web_handler=731ns query_engine=7.204µs scrape=48.411µs scrape_sd=1.062µs notify=1.022µs notify_sd=471ns rules=232.977µs tracing=4.859µs
            ts=2024-10-02T00:51:31.066Z caller=main.go:1399 level=info msg="Loading configuration file" filename=/tmp/config654002981
            ts=2024-10-02T00:51:31.066Z caller=main.go:1449 level=info msg="Completed loading of configuration file" filename=/tmp/config654002981 totalDuration=506.4µs db_storage=1.844µs remote_storage=1.804µs web_handler=541ns query_engine=17.333µs scrape=95.099µs scrape_sd=1.292µs notify=641ns notify_sd=581ns rules=192.06µs tracing=5.57µs
            
ts=2024-10-02T00:52:04.597Z caller=main.go:601 level=info msg="No time or size retention was set so using the default time retention" duration=15d
ts=2024-10-02T00:52:04.597Z caller=main.go:645 level=info msg="Starting Prometheus Server" mode=server version="(version=, branch=, revision=unknown)"
ts=2024-10-02T00:52:04.597Z caller=main.go:650 level=info build_context="(go=go1.23.1, platform=linux/amd64, user=, date=, tags=dedupelabels)"
ts=2024-10-02T00:52:04.597Z caller=main.go:651 level=info host_details="(Linux 6.8.0-1014-azure #16~22.04.1-Ubuntu SMP Thu Aug 15 21:31:41 UTC 2024 x86_64 10bad815c49d (none))"
ts=2024-10-02T00:52:04.597Z caller=main.go:652 level=info fd_limits="(soft=1048576, hard=1048576)"
ts=2024-10-02T00:52:04.597Z caller=main.go:653 level=info vm_limits="(soft=unlimited, hard=unlimited)"
ts=2024-10-02T00:52:04.600Z caller=web.go:651 level=info component=web msg="Start listening for connections" address=127.0.0.1:43889
ts=2024-10-02T00:52:04.601Z caller=main.go:1192 level=info msg="Starting TSDB ..."
ts=2024-10-02T00:52:04.603Z caller=tls_config.go:348 level=info component=web msg="Listening on" address=127.0.0.1:43889
ts=2024-10-02T00:52:04.603Z caller=tls_config.go:351 level=info component=web msg="TLS is disabled." http2=false address=127.0.0.1:43889
ts=2024-10-02T00:52:04.608Z caller=head.go:632 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
ts=2024-10-02T00:52:04.608Z caller=head.go:719 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=1.072µs
ts=2024-10-02T00:52:04.608Z caller=head.go:727 level=info component=tsdb msg="Replaying WAL, this may take a while"
ts=2024-10-02T00:52:04.608Z caller=head.go:799 level=info component=tsdb msg="WAL segment loaded" segment=0 maxSegment=0
ts=2024-10-02T00:52:04.608Z caller=head.go:836 level=info component=tsdb msg="WAL replay completed" checkpoint_replay_duration=72.637µs wal_replay_duration=141.495µs wbl_replay_duration=131ns chunk_snapshot_load_duration=0s mmap_chunk_replay_duration=1.072µs total_replay_duration=232.336µs
ts=2024-10-02T00:52:04.610Z caller=main.go:1213 level=info fs_type=794c7630
ts=2024-10-02T00:52:04.610Z caller=main.go:1216 level=info msg="TSDB started"
ts=2024-10-02T00:52:04.610Z caller=main.go:1399 level=info msg="Loading configuration file" filename=/tmp/TestScrapeFailureLogFile2971671619/001/prometheus.yml
ts=2024-10-02T00:52:04.610Z caller=main.go:1438 level=info msg="updated GOGC" old=100 new=75
ts=2024-10-02T00:52:04.610Z caller=main.go:1449 level=info msg="Completed loading of configuration file" filename=/tmp/TestScrapeFailureLogFile2971671619/001/prometheus.yml totalDuration=739.337µs db_storage=1.002µs remote_storage=1.[223](https://github.com/prometheus/prometheus/actions/runs/11135242372/job/30944893362?pr=15068#step:6:224)µs web_handler=441ns query_engine=652ns scrape=384.501µs scrape_sd=23.835µs notify=792ns notify_sd=571ns rules=1.022µs tracing=5.931µs
ts=2024-10-02T00:52:04.610Z caller=main.go:1177 level=info msg="Server is ready to receive web requests."
ts=2024-10-02T00:52:04.611Z caller=manager.go:164 level=info component="rule manager" msg="Starting rule manager..."
ts=2024-10-02T00:52:10.571Z caller=main.go:1399 level=info msg="Loading configuration file" filename=/tmp/TestScrapeFailureLogFile2971671619/001/prometheus.yml
ts=2024-10-02T00:52:10.571Z caller=main.go:1449 level=info msg="Completed loading of configuration file" filename=/tmp/TestScrapeFailureLogFile2971671619/001/prometheus.yml totalDuration=615.445µs db_storage=1.423µs remote_storage=1.803µs web_handler=541ns query_engine=932ns scrape=329.528µs scrape_sd=14.898µs notify=811ns notify_sd=1.071µs rules=2.004µs tracing=5.33µs
ts=2024-10-02T00:52:11.573Z caller=main.go:1399 level=info msg="Loading configuration file" filename=/tmp/TestScrapeFailureLogFile2971671619/001/prometheus.yml
ts=2024-10-02T00:52:11.573Z caller=main.go:1449 level=info msg="Completed loading of configuration file" filename=/tmp/TestScrapeFailureLogFile2971671619/001/prometheus.yml totalDuration=395.641µs db_storage=1.383µs remote_storage=1.442µs web_handler=471ns query_engine=1.012µs scrape=197.28µs scrape_sd=12.773µs notify=1.202µs notify_sd=682ns rules=1.603µs tracing=5.54µs
FAIL
FAIL	github.com/prometheus/prometheus/cmd/prometheus	111.685s
```

### System information

_No response_

### Prometheus version

_No response_

### Prometheus configuration file

_No response_

### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

--- a/cmd/prometheus/query_log_test.go
+++ b/cmd/prometheus/query_log_test.go
@@ -125,10 +125,59 @@ func (p *queryLogTest) query(t *testing.T) {
 		require.NoError(t, err)
 		require.Equal(t, 200, r.StatusCode)
 	case ruleOrigin:
-		time.Sleep(2 * time.Second)
+		// Poll the /api/v1/rules endpoint until a new rule evaluation is detected.
+		var lastEvalTime time.Time
+		for {
+			r, err := http.Get(fmt.Sprintf("http://%s:%d/api/v1/rules", p.host, p.port))
+			require.NoError(t, err)
+
+			rulesBody, err := io.ReadAll(r.Body)
+			require.NoError(t, err)
+			defer r.Body.Close()
+
+			// Parse the rules response to find the last evaluation time.
+			newEvalTime := parseLastEvaluation(rulesBody)
+			if newEvalTime.After(lastEvalTime) {
+				if !lastEvalTime.IsZero() {
+					break
+				}
+				lastEvalTime = newEvalTime
+			}
+
+			time.Sleep(100 * time.Millisecond)
+		}
 	default:
 		panic("can't query this origin")
 	}
+}
+
+// parseLastEvaluation extracts the last evaluation timestamp from the /api/v1/rules response.
+func parseLastEvaluation(rulesBody []byte) time.Time {
+	var ruleResponse struct {
+		Status string `json:"status"`
+		Data   struct {
+			Groups []struct {
+				Rules []struct {
+					LastEvaluation string `json:"lastEvaluation"`
+				} `json:"rules"`
+			} `json:"groups"`
+		} `json:"data"`
+	}
+
+	err := json.Unmarshal(rulesBody, &ruleResponse)
+	if err != nil {
+		return time.Time{}
+	}
+
+	for _, group := range ruleResponse.Data.Groups {
+		for _, rule := range group.Rules {
+			if evalTime, err := time.Parse(time.RFC3339Nano, rule.LastEvaluation); err == nil {
+				return evalTime
+			}
+		}
+	}
+
+	return time.Time{}
 }
 
 // queryString returns the expected queryString of a this test.
@@ -322,7 +371,7 @@ func (p *queryLogTest) run(t *testing.T) {
 	if p.exactQueryCount() {
 		require.Len(t, ql, qc)
 	} else {
-		require.Greater(t, len(ql), qc, "no queries logged")
+		require.GreaterOrEqual(t, len(ql), qc, "no queries logged")
 	}
 	p.validateLastQuery(t, ql)
 	qc = len(ql)
@@ -353,7 +402,7 @@ func (p *queryLogTest) run(t *testing.T) {
 	if p.exactQueryCount() {
 		require.Len(t, ql, qc)
 	} else {
-		require.Greater(t, len(ql), qc, "no queries logged")
+		require.GreaterOrEqual(t, len(ql), qc, "no queries logged")
 	}
 	p.validateLastQuery(t, ql)
 

